### PR TITLE
[e-m-c][ios] Expose `NativeModulesProxy` config as a JSI module

### PR DESCRIPTION
**Pull request by barthap**
# Why

Resolves ENG-5091

# How

> I made a few iterations and experiments with several approaches - this PR is a proposal of an approach which is pretty fast, clean and not that invasive.

- Created the `NativeProxyModule` - a Sweet expo-module which is a counterpart of the `NativeUnimoduleProxy` RN module (`EXNativeModulesProxy`)
  - it is registered manually in `ExpoBridgeModule`, right after the creation of `AppContext`
  - currently, it only exports constants - the same as the `EXNativeModulesProxy` module
  - the constants creation logic is still written in Objective-C inside `EXNativeModulesProxy`. It will finally be migrated in the future, but there's no reason to move it in this PR.
- Refactored config creation in `EXNativeModulesProxy constantsToExport`
  - created the `EXModulesProxyConfig` class - a store for the module proxy config + some convenience methods for dictionary management
  - config for both legacy modules and new sweet expo-modules is now lazy-loaded and cached
  - minimized Swift calls from Objective-C, as they were performance bottlenecks - now there is only one such call.
- Yet, both configs are loaded in `NativeModulesProxy.native.ts` - config constants are taken from the JSI module first (with fallback to legacy constants), but method calls are still handled by the legacy proxy module.
  - We need to load both (or at least the legacy one) anyway, but this shouldn't be a big issue for now, because the config is created only once and cached.

# Test Plan

Speed comparison: faster than the legacy proxy - usually noted `8ms` vs `12ms`

- NCL on bare-expo runs
- The Expo Modules screen on NCL displays all config as constants (will need to edit that screen, it is too long)

### CODE REVIEW

File: packages/expo-modules-core/ios/NativeModulesProxy/NativeProxyModule.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+public class NativeProxyModule: Module {
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/NativeModulesProxy/NativeProxyModule.swift:3]: Actually I would rename it to `NativeModulesProxyModule` 🤔

File: packages/expo-modules-core/ios/Swift/AppContext.swift [hunk 2]
```diff
@@ -31,6 +31,9 @@ public final class AppContext: NSObject {
   @objc
   public weak var legacyModuleRegistry: EXModuleRegistry?
 
+  @objc
+  public weak var legacyModulesProxy: LegacyNativeModulesProxy?
```
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-modules-core/ios/Swift/AppContext.swift:35]: Did you mean _"ask legacy proxy for the legacy registry"_? as actually, the proxy is what holds a reference to the registry, not the opposite.
Or is `EXNativeModulesProxy` also registered as a module in the registry (I don't think so but not sure) and you mean to get it using one of the registry methods?

If you meant the first option, I think I tried to do it before, I'm not sure why I gave up.

[the remark this answers is not in the document]
Comment (tsapeta) [reply packages/expo-modules-core/ios/Swift/AppContext.swift:35]: I meant the second option, but you're probably right that we don't put it into the registry (it's just a bridge module)

File: packages/expo-modules-core/src/NativeModulesProxy.native.ts [hunk 3]
```diff
@@ -23,7 +31,9 @@ if (NativeProxy) {
             )
           );
         }
-        return NativeProxy.callMethod(moduleName, key, args);
+
+        // we still want to call methods using the legacy proxy
```
Comment (tsapeta) [inline packages/expo-modules-core/src/NativeModulesProxy.native.ts:35]:
```suggestion
        // We still want to call methods using the legacy proxy in SDK46
```

File: packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.mm [hunk 4]
```diff
@@ -28,6 +28,47 @@
...
+}
+
+- (void)addEntriesFromConfig:(const EXModulesProxyConfig*)config
+{
+  [_exportedConstants addEntriesFromDictionary:config.exportedConstants];
+  [_methodNames addEntriesFromDictionary:config.methodNames];
+  [_viewManagerMetadata addEntriesFromDictionary:config.viewManagerMetadata];
+}
+
+- (NSDictionary<NSString*, id> *) toDictionary
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.mm:61]:
```suggestion
- (nonnull NSDictionary<NSString *, id> *)toDictionary
```

File: packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.h [hunk 5]
```diff
@@ -5,13 +5,29 @@
...
+
+NS_SWIFT_NAME(ModulesProxyConfig)
+@interface EXModulesProxyConfig : NSObject
+
+- (instancetype)initWithConstants:(nonnull NSDictionary *)constants
+                      methodNames:(nonnull NSDictionary *)methodNames
+                     viewManagers:(nonnull NSDictionary *)viewManagerMetadata;
+
+- (void)addEntriesFromConfig:(nonnull const EXModulesProxyConfig *)config;
+- (NSDictionary<NSString *, id> *) toDictionary;
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.h:18]:
```suggestion
- (nonnull NSDictionary<NSString *, id> *)toDictionary;
```

File: packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.mm [hunk 6]
```diff
@@ -28,6 +28,47 @@
...
+  self = [super init];
+  if (self) {
+    _exportedConstants = constants;
+    _methodNames = methodNames;
+    _viewManagerMetadata = viewManagerMetadata;
+  }
+  return self;
+}
+
+- (void)addEntriesFromConfig:(const EXModulesProxyConfig*)config
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.mm:54]:
```suggestion
- (void)addEntriesFromConfig:(nonnull const EXModulesProxyConfig*)config
```

File: packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.mm [hunk 7]
```diff
@@ -28,6 +28,47 @@
...
+@end
+
+@implementation EXModulesProxyConfig
+
+- (instancetype)initWithConstants:(nonnull NSDictionary *)constants
+                      methodNames:(nonnull NSDictionary *)methodNames
+                     viewManagers:(nonnull NSDictionary *)viewManagerMetadata
+{
+  self = [super init];
+  if (self) {
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/NativeModulesProxy/EXNativeModulesProxy.mm:46]:
```suggestion
  if (self = [super init]) {
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**